docs/proposals/client-package-structure.md

```diff
@@ -0,0 +1,227 @@
+# Client: layering and package structure
+
+## Desired layers
+
+### Transport
+
+The transport layer is concerned with round-tripping requests to an apiserver
+somewhere. It consumes a Config object with options appropriate for this.
+(That's most of the current client.Config structure.)
+
+Transport delivers an object that implements http's RoundTripper interface
+and/or can be used in place of http.DefaultClient to route requests.
+
+Transport objects are safe for concurrent use, and are cached and reused by
+subsequent layers.
+
+Tentative name: "Transport".
+
+It's expected that the transport config will be general enough that third
+parties (e.g., OpenShift) will not need their own implementation, rather they
+can change the Host field to point at their own servers, etc..
```

Comment:
Looks good to me.

On Tue, Oct 20, 2015 at 5:35 PM, krousey <notifications@github.com> wrote:

> In docs/proposals/client-package-structure.md
> <https://github.com/kubernetes/kubernetes/pull/15730#discussion_r42557856>
> :
>
> > +
> > +The transport layer is concerned with round-tripping requests to an apiserver
> > +somewhere. It consumes a Config object with options appropriate for this.
> > +(That's most of the current client.Config structure.)
> > +
> > +Transport delivers an object that implements http's RoundTripper interface
> > +and/or can be used in place of http.DefaultClient to route requests.
> > +
> > +Transport objects are safe for concurrent use, and are cached and reused by
> > +subsequent layers.
> > +
> > +Tentative name: "Transport".
> > +
> > +It's expected that the transport config will be general enough that third
> > +parties (e.g., OpenShift) will not need their own implementation, rather they
> > +can change the Host field to point at their own servers, etc..
>
> I'm actually not seeing anything in the transport layer that is specific
> to hosts. That all seems to be handled at the request layer.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/15730/files#r42557856>.
>